internal/provisioner/prometheus.go

```diff
@@ -0,0 +1,110 @@
+package provisioner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/mattermost/mattermost-cloud/internal/tools/aws"
+	"github.com/mattermost/mattermost-cloud/internal/tools/kops"
+	"github.com/mattermost/mattermost-cloud/model"
+	"github.com/pborman/uuid"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type prometheus struct {
+	awsClient   aws.AWS
+	cluster     *model.Cluster
+	kops        *kops.Cmd
+	logger      log.FieldLogger
+	provisioner *KopsProvisioner
+}
+
+func newPrometheusHandle(cluster *model.Cluster, provisioner *KopsProvisioner, awsClient aws.AWS, kops *kops.Cmd, logger log.FieldLogger) (*prometheus, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("cannot instantiate Prometheus handle with nil logger")
+	}
+
+	if cluster == nil {
+		return nil, fmt.Errorf("Cannot create a connection to Prometheus is the cluster provided is nil")
+	}
+
+	if provisioner == nil {
+		return nil, fmt.Errorf("Cannot create a connection to Prometheus is the provisioner provided is nil")
+	}
+
+	if awsClient == nil {
+		return nil, fmt.Errorf("Cannot create a connection to Prometheus is the awsClient provided is nil")
+	}
+
+	if kops == nil {
+		return nil, fmt.Errorf("Cannot create a connection to Prometheus is the Kops command provided is nil")
+	}
+
+	return &prometheus{
+		cluster:     cluster,
+		provisioner: provisioner,
+		awsClient:   awsClient,
+		kops:        kops,
+		logger:      logger.WithField("prometheus-handle", uuid.New()),
+	}, nil
+}
+
+func (p *prometheus) Create() error {
+	err := p.NewHelmDeployment().Create()
+	if err != nil {
+		return err
+	}
+
+	logger := p.logger.WithField("prometheus-create", uuid.New())
+
+	app := "prometheus"
+	dns := fmt.Sprintf("%s.%s.%s", p.cluster.ID, app, p.provisioner.privateDNS)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 120)
+	defer cancel()
+
+	endpoint, err := getLoadBalancerEndpoint(ctx, "internal-nginx", logger.WithField("prometheus-create", uuid.New()), p.kops.GetKubeConfigPath())
```

Comment:
My main concern now is that although it definitely makes sense to add new logger fields for extra info, I think this may be a bit too much. By this time line 67 is called here we have three additional UUIDs used when logging only prometheus create actions.

Considering we will probably have other fields including cluster ID, I would probably leave out the UUIDs and do something like this:

```
	return &prometheus{
		cluster:     cluster,
		provisioner: provisioner,
		awsClient:   awsClient,
		kops:        kops,
		logger:      logger.WithField("cluster-utility", "prometheus"),
	}, nil

...

func (p *prometheus) Create() error {
	err := p.NewHelmDeployment().Create()
	if err != nil {
		return err
	}

	logger := p.logger.WithField("prometheus-action", "create")
...
```

I am not totally against UUIDs here, but I would only add one if any. Keep in mind that when doing any main provisioner action we already have a `run` UUID that would be passed in here.

```
time="2020-01-07T14:54:13Z" level=warning msg="[kops] terraform plan" cluster=og4nzpbstbrszbdwynf9zxza4o instance=6mt635wzqpbnunhsu574iragqc run=jee3iyhbz7n4t8bae1yyhbd66a
time="2020-01-07T14:54:13Z" level=warning msg="[kops] terraform apply" cluster=og4nzpbstbrszbdwynf9zxza4o instance=6mt635wzqpbnunhsu574iragqc run=jee3iyhbz7n4t8bae1yyhbd66a
time="2020-01-07T14:54:13Z" level=warning msg="[kops] Suggestions:" cluster=og4nzpbstbrszbdwynf9zxza4o instance=6mt635wzqpbnunhsu574iragqc run=jee3iyhbz7n4t8bae1yyhbd66a
```

Reply:
Yeah, I was torn on this. I thought maybe it would be nice to be able to trace these calls,  but it's definitely overkill for the current state of the design and there is likely a runtime cost for generating the UUIDs. But I thought just providing a static string would be redundant. I'll switch to a static string since I feel mixed about both options and you seem to prefer that.